test/functional/mining.py

```diff
@@ -0,0 +1,336 @@
+      when creating blocks.
+    - Test that recently received transactions aren't selected unless the feerate
+      for including them is high.
+'''
+
+class MiningTest(BitcoinTestFramework):
+    def __init__(self):
+        self.num_nodes = 6
+        self.setup_clean_chain = True
+
+    def setup_network(self, split=False):
+        # Set blockmaxweight to be low, to require fewer transactions
+        # to fill up a block.
+        self.cnb_args = [ { 'window': 10000, 'stalerate': 0.01},
+                          { 'window': 30000, 'stalerate': 0.1},
+                          { 'window': 50000, 'stalerate': 0.001}
+                        ]
+
+        extra_args = [["-debug", "-blockmaxweight=400000", "-minrelaytxfee=0"]
+                      for i in range(self.num_nodes)]
+        for i in range(3):
+            extra_args[i].extend(["-recenttxwindow=%s"%self.cnb_args[i]['window'],
+                                  "-recenttxstalerate=%s"%self.cnb_args[i]['stalerate']])
+        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, extra_args)
+
+        # Connect the network in a loop
+        for i in range(self.num_nodes-1):
+            connect_nodes(self.nodes[i], i+1)
+        connect_nodes(self.nodes[-1], 0)
+
+        self.is_network_split = False
+
+    def create_segwit_coins(self):
+        for node in self.nodes:
+            # Create 20 segwit coins on each node
+            for i in range(20):
+                new_addr = node.getnewaddress()
+                witness_address = node.addwitnessaddress(new_addr)
+                node.sendtoaddress(witness_address, 1)
+            node.generate(1)
+            self.sync_all()
+
+    def populate_mempool(self, desired_transactions):
+        num_transactions = 0
+        while (num_transactions < desired_transactions):
+            try:
+                random_amount = random.randint(1, 10)*Decimal('0.1')
+                random_transaction(self.nodes, amount=random_amount, min_fee=Decimal("0.00001"), fee_increment=Decimal("0.000005"), fee_variants=5000, confirmations_required=0)
+                num_transactions += 1
+            except RuntimeError as e:
+                # We might run out of funds; just count these as valid attempts
+                if ("Insufficient funds" in e.error['message']):
+                    num_transactions += 1
+                else:
+                    raise AssertionError("Unexpected run time error: "+e.error['message'])
+            except JSONRPCException as e:
+                if ("too-long-mempool" in e.error['message']):
+                    num_transactions += 1
+                else:
+                    raise AssertionError("Unexpected JSON-RPC error: "+e.error['message'])
+            except Exception as e:
+                raise AssertionError("Unexpected exception raised: "+type(e).__name__)
+        self.sync_all()
+
+    # Approximate weight of the transactions in the mempool
```

Comment:
In commit "Add functional test for CreateNewBlock"

Some of the comments on top these python functions could be made into docstrings.